sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala

```diff
@@ -353,6 +354,59 @@ trait CheckAnalysis extends PredicateHelper {
               case _ =>
             }
 
+          case alter: AlterTable if alter.childrenResolved =>
```

Comment:
Is there a reason that some of these checks are here, and some in the `resolved` of AlterTable? It'd be nice to hide the complexity inside AlterTable if possible, and do a simpler check here on whether AlterTable is `resolved`.
However, I would much prefer the case where we can throw a better error message.

Reply:
Yes. The meaning of `resolved` is that references are satisfied and there are no unresolved placeholders -- that's why `resolved` checks that the referenced fields exist. This checks whether that resolved plan has valid and reasonable changes, like whether the change updates a column from a string to a boolean.